btx/processing/freqdir.py

```diff
@@ -0,0 +1,740 @@
+import os, csv, argparse
+
+import numpy as np
+from mpi4py import MPI
+
+from matplotlib import pyplot as plt
+from matplotlib import colors
+
+from btx.misc.shortcuts import TaskTimer
+
+from btx.interfaces.ipsana import (
+    PsanaInterface,
+    bin_data,
+    bin_pixel_index_map,
+    retrieve_pixel_index_map,
+    assemble_image_stack_batch,
+)
+
+###########################################
+#John Imports
+from numpy import zeros, sqrt, dot, diag
+from numpy.linalg import svd, LinAlgError
+from scipy.linalg import svd as scipy_svd
+import numpy as np
+
+import time
+
+from datetime import datetime
+currRun = datetime.now().strftime("%y%m%d%H%M%S")
+
+import h5py
+
+from PIL import Image
+
+#writeDirec = "h5writes/"
+#############################################
+
+class FreqDir:
+
+    """Parallel Frequent Directions."""
+
+    def __init__(
+        self,
+        john_start,
+        tot_imgs,
+        ell, 
+        alpha,
+        exp,
+        run,
+        det_type,
+        rankAdapt,
+        writeDirec,
+        merger=False,
+        mergerFeatures=0,
+        downsample=False,
+        bin_factor=2,
+        output_dir=""
+    ):
+
+        self.writeDirec = writeDirec
+
+        self.comm = MPI.COMM_WORLD
+        self.rank = self.comm.Get_rank()
+        self.size = self.comm.Get_size()
+        self.merger = merger
+
+        if not merger:
+            self.psi = PsanaInterface(exp=exp, run=run, det_type=det_type)
+            self.psi.counter = john_start + tot_imgs*self.rank//self.size
+
+            self.downsample = downsample
+            self.bin_factor = bin_factor
+            self.output_dir = output_dir
+
+            (
+                self.num_images,
+                _,
+                self.num_features,
+            ) = self.set_params(tot_imgs, ell, bin_factor)
```

Comment:
Do you think we could envision having `FreqDir` and `PiPCA` share some features? Maybe by introducing a base class that both inherit from? This would greatly help maintain both in my opnion.